### ISSUE TITLE
Unused recursive function is not reported by clang-tidy

### ISSUE THREAD

**Issue by vskod** [llvm/llvm-project#60838]
Here is source code to reproduce the issue:
```
int sum(int v)
{
    return v > 0 ? v + sum(v - 1) : 0;
}

int main()
{
}
```

**Comment by DimitryAndric** [comment]
I don't think there is anything wrong? Because `sum()` is not a static function, it could be called from other objects, and therefore it would be incorrect to remove it. If you declare the function static, it should complain.

**Comment by vskod** [comment]
Ok, here is slightly modified code:
```c++
static int sum(int v)
{
    return v > 0 ? v + sum(v - 1) : 0;
}

static int zero()
{
    return 0;
}

int main()
{
}
```
and clang-tidy output:
```bash
$ clang-tidy -p cmake-build-debug/compile_commands.json unused.cpp 
2 warnings generated.
/build/work/playground/unused.cpp:1:12: warning: function 'sum' is not needed and will not be emitted [clang-diagnostic-unneeded-internal-declaration]
static int sum(int v)
           ^
/build/work/playground/unused.cpp:6:12: warning: unused function 'zero' [clang-diagnostic-unused-function]
static int zero()
           ^
```
Those warnings are different.

**Comment by firewave** [comment]
@DimitryAndric is right. If you enable all warnings it will tell you. But there's a case that would split through:

```cpp
int sum1(int v)
{
    return v > 0 ? v + sum1(v - 1) : 0;
}

static int sum2(int v)
{
    return v > 0 ? v + sum2(v - 1) : 0;
}

inline int sum3(int v)
{
    return v > 0 ? v + sum3(v - 1) : 0;
}

static inline int sum4(int v)
{
    return v > 0 ? v + sum4(v - 1) : 0;
}

int main()
{
}
```

```
<source>:1:5: warning: no previous prototype for function 'sum1' [-Wmissing-prototypes]
int sum1(int v)
    ^
<source>:1:1: note: declare 'static' if the function is not intended to be used outside of this translation unit
int sum1(int v)
^
static 
<source>:6:12: warning: function 'sum2' is not needed and will not be emitted [-Wunneeded-internal-declaration]
static int sum2(int v)
           ^
<source>:16:19: warning: function 'sum4' is not needed and will not be emitted [-Wunneeded-internal-declaration]
static inline int sum4(int v)
                  ^
```
https://godbolt.org/z/WK9Krn15x

There's no `-Wmissing-prototypes` for `inline` functions and thus if those are actually just internal and unused they will never be reported.
